### PR TITLE
community[minor]: add UpTrainCallbackHandler

### PR DESCRIPTION
**Pull request by Dominastorm**
- **Description:** 
This PR adds a callback handler for UpTrain. It performs evaluations in the RAG pipeline to check the quality of retrieved documents, generated queries and responses. 

- **Dependencies:** 
    - The UpTrainCallbackHandler requires the uptrain package

### CODE REVIEW

File: libs/community/langchain_community/callbacks/uptrain_callback.py [hunk 1]
```diff
@@ -0,0 +1,390 @@
+"""UpTrain's Callback Handler."""
+#Langchain Community
```
Comment (eyurtsev) [inline libs/community/langchain_community/callbacks/uptrain_callback.py:2]:
```suggestion
```

File: libs/community/langchain_community/callbacks/uptrain_callback.py [hunk 2]
```diff
@@ -0,0 +1,390 @@
+"""UpTrain's Callback Handler."""
```
Comment (eyurtsev) [inline libs/community/langchain_community/callbacks/uptrain_callback.py:1]: Consider expanding module documentation a bit (e.g., include link to uptrain etc.)

File: libs/community/langchain_community/callbacks/uptrain_callback.py [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+from uuid import UUID
+
+from langchain.schema import AgentAction, AgentFinish, LLMResult
+from langchain_core.callbacks.base import AsyncCallbackHandler
+from langchain_core.documents import Document
+from langchain_core.messages.base import BaseMessage
+from uptrain import APIClient, EvalLLM, Evals, Settings
+
+
+def import_uptrain():
```
Comment (eyurtsev) [inline libs/community/langchain_community/callbacks/uptrain_callback.py:24]: missing type information

File: libs/community/langchain_community/callbacks/uptrain_callback.py [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+
+    Args:
+        BaseCallbackHandler (_type_): _description_
+    """
+
+    REPO_URL: str = "https://github.com/uptrain-ai/uptrain"
+    ISSUES_URL: str = f"{REPO_URL}/issues"
+    DOCS_URL: str = "https://docs.uptrain.ai"
+
+    import os
```
Comment (eyurtsev) [inline libs/community/langchain_community/callbacks/uptrain_callback.py:75]:
```suggestion
```

File: libs/community/langchain_community/callbacks/uptrain_callback.py [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+            project_name_prefix: str = "langchain",
+            key_type = "openai",
+            api_key = os.getenv("OPENAI_API_KEY"),
+            ) -> None:
+        """Initializes the `UpTrainCallbackHandler`."""
+        super().__init__()
+
+        uptrain = import_uptrain()
+
+        import os
```
Comment (eyurtsev) [inline libs/community/langchain_community/callbacks/uptrain_callback.py:87]: move to global scope

File: libs/community/langchain_community/callbacks/uptrain_callback.py [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+    REPO_URL: str = "https://github.com/uptrain-ai/uptrain"
+    ISSUES_URL: str = f"{REPO_URL}/issues"
+    DOCS_URL: str = "https://docs.uptrain.ai"
+
+    import os
+    def __init__(
+            self,
+            project_name_prefix: str = "langchain",
+            key_type = "openai",
+            api_key = os.getenv("OPENAI_API_KEY"),
```
Comment (eyurtsev) [inline libs/community/langchain_community/callbacks/uptrain_callback.py:80]: why are we picking up openapi keys from the env?

File: libs/community/langchain_community/callbacks/uptrain_callback.py [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+            "score_context_reranking": "Context Reranking Score",
+            "score_context_conciseness": "Context Conciseness Score",
+            "score_multi_query_accuracy": "Multi Query Accuracy Score",
+        }
+
+        for row in uptrain_result:
+            columns = list(row.keys())
+            for column in columns:
+                if column == "question":
+                    print(f"\nQuestion: {row[column]}")
```
Comment (eyurtsev) [inline libs/community/langchain_community/callbacks/uptrain_callback.py:141]: Why are there print statements here?

File: libs/community/langchain_community/callbacks/uptrain_callback.py [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+                    if not self.first_score_printed_flag:
+                        print()
+                        self.first_score_printed_flag = True
+                    if column in score_name_map:
+                        print(f"{score_name_map[column]}: {row[column]}")
+                    else:
+                        print(f"{column}: {row[column]}")
+            print()
+
+    def on_llm_start(
```
Comment (eyurtsev) [inline libs/community/langchain_community/callbacks/uptrain_callback.py:161]: The sync methods should not exist on this callback handler. The callback handler inherits from an async callback handler and should only implement the async methods that are needed

**Review by eyurtsev — CHANGES_REQUESTED** [review]

File: libs/community/langchain_community/callbacks/uptrain_callback.py [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+        self.multi_query_daugher_run_id: str = ""
+
+class UpTrainCallbackHandler(AsyncCallbackHandler):
+    """Callback Handler that logs to uptrain.
+
+    Args:
+        BaseCallbackHandler (_type_): _description_
+    """
+
+    REPO_URL: str = "https://github.com/uptrain-ai/uptrain"
```
Comment (eyurtsev) [inline libs/community/langchain_community/callbacks/uptrain_callback.py:71]: Why is this here?

File: libs/community/langchain_community/callbacks/uptrain_callback.py (hunk 6 above)
Comment (Dominastorm) [reply libs/community/langchain_community/callbacks/uptrain_callback.py:79]: My apologies, I had it for testing and missed it while reviewing. I have removed it now.

File: libs/community/langchain_community/callbacks/uptrain_callback.py (hunk 8 above)
Comment (Dominastorm) [reply libs/community/langchain_community/callbacks/uptrain_callback.py:160]: Changed it to BaseCallbackHandler

File: libs/community/langchain_community/callbacks/uptrain_callback.py (hunk 9 above)
Comment (Dominastorm) [reply libs/community/langchain_community/callbacks/uptrain_callback.py]: I had used https://github.com/langchain-ai/langchain/blob/master/libs/community/langchain_community/callbacks/argilla_callback.py for reference. I see that other callback files don't have it. I have removed it now.

File: libs/community/langchain_community/callbacks/uptrain_callback.py (hunk 2 above)
Comment (Dominastorm) [reply libs/community/langchain_community/callbacks/uptrain_callback.py]: Thanks for the sugeestion. I have done the same.

File: libs/community/langchain_community/callbacks/uptrain_callback.py (hunk 7 above)
Comment (Dominastorm) [reply libs/community/langchain_community/callbacks/uptrain_callback.py:177]: The intention is that when a user uses the uptrain callback handler and calls a retriever, information is shown in the following format without the user needing to do any additional setup:

image


Do you recommend I use logger or another alternative instead?

File: libs/langchain/langchain/callbacks/__init__.py [hunk 10]
```diff
@@ -79,4 +79,5 @@ def __getattr__(name: str) -> Any:
     "SageMakerCallbackHandler",
     "LabelStudioCallbackHandler",
     "TrubricsCallbackHandler",
+    "UpTrainCallbackHanlder",
```
Comment (eyurtsev) [inline libs/langchain/langchain/callbacks/__init__.py:82]: We are in the process of breaking dependency of langchain on community, so we don't want anyone to be importing community based code from `langchain`

```suggestion
```

File: libs/community/langchain_community/callbacks/uptrain_callback.py [hunk 11]
```diff
@@ -0,0 +1,425 @@
+"""
```
Comment (eyurtsev) [inline libs/community/langchain_community/callbacks/uptrain_callback.py:1]: Nice

File: libs/community/langchain_community/callbacks/uptrain_callback.py [hunk 12]
```diff
@@ -0,0 +1,425 @@
...
+        self.multi_query_daugher_run_id: str = ""
+
+
+class UpTrainCallbackHandler(BaseCallbackHandler):
+    """Callback Handler that logs evalution results to uptrain and the console.
+
+    Args:
+        project_name_prefix (str): Prefix for the project name.
+        key_type (str): Type of key to use. Must be 'uptrain' or 'openai'.
+        api_key (str): API key for the UpTrain or OpenAI API.
```
Comment (eyurtsev) [inline libs/community/langchain_community/callbacks/uptrain_callback.py:110]: Why is an openai API key required? Could you document what it's used for?

**Review by eyurtsev — COMMENTED** [review]
OK almost ready to merge:

1) Could you clarify purpose of open ai key in the doc-strings for the class?
2) Remove methods with empty implementation if possible (if I'm not mistaken base class makes those optional already)
3) remove imports in langchain pkg from langchain_community

For the print you could accept a logger function and default to print or configure a logger. If you're configuring a logger I'd just check that output is visible in an interactive environment like jupyter

File: libs/community/langchain_community/callbacks/uptrain_callback.py (hunk 11 above)
Comment (Dominastorm) [reply libs/community/langchain_community/callbacks/uptrain_callback.py:1]: Thanks!

File: libs/community/langchain_community/callbacks/uptrain_callback.py (hunk 12 above)
Comment (Dominastorm) [reply libs/community/langchain_community/callbacks/uptrain_callback.py:110]: LLM calls are used to perform the evaluations

File: libs/community/langchain_community/callbacks/uptrain_callback.py [hunk 13]
```diff
@@ -0,0 +1,395 @@
...
+    Union,
+)
+from uuid import UUID
+
+from langchain.schema import AgentAction, AgentFinish, LLMResult
+from langchain_core.callbacks.base import BaseCallbackHandler
+from langchain_core.documents import Document
+from langchain_core.messages.base import BaseMessage
+
+logging.basicConfig(format='%(message)s', stream=sys.stdout)
```
Comment (eyurtsev) [inline libs/community/langchain_community/callbacks/uptrain_callback.py:63]: I believe this changes the logging configuration globally -- which we shouldn't do

Options: 
1) Either configure the logger in this file to output to sys stdout.
2) Accept a function from a user for the logs (and maybe default it to the print function?)
3) Revert back to print if you can't find a way to do (1-2) -- before doing that I'd just check that there's a way to silence the linter. In some sense `prints` are OK for this since this is explicitly for logging purposes

File: libs/community/langchain_community/callbacks/uptrain_callback.py [hunk 14]
```diff
@@ -0,0 +1,395 @@
...
+            self.uptrain_client = uptrain.EvalLLM(settings=settings)
+        else:
+            raise ValueError("Invalid key type: Must be 'uptrain' or 'openai'")
+
+    def uptrain_evaluate(
+        self,
+        project_name: str,
+        data: List[Dict[str, str]],
+        checks: List[str],
+        log_results: bool = True,
```
Comment (eyurtsev) [inline libs/community/langchain_community/callbacks/uptrain_callback.py:173]:
```suggestion
        *,
        log_results: bool = True,
```

File: libs/community/langchain_community/callbacks/uptrain_callback.py (hunk 13 above)
Comment (Dominastorm) [reply libs/community/langchain_community/callbacks/uptrain_callback.py]: I have modified it, is it fine now?

**Review by eyurtsev — APPROVED** [review]

**Comment by Dominastorm** [comment]
Hey @eyurtsev, I am done with all the changes from my side. Is there anything else that needs to be done?

**Review by baskaryan — APPROVED** [review]

**Comment by eyurtsev** [comment]
Apologies for delay -- set on auto merge last time and didn't realize this didn't get merged.

**Review by eyurtsev — APPROVED** [review]

**Comment by Dominastorm** [comment]
@eyurtsev, sorry to make you do this so many times. I ran the `make lint` command locally, but it hadn't given the same errors. I have made appropriate fixes now. Can you approve it? Hopefully, it will pass this time

**Comment by eyurtsev** [comment]
@Dominastorm no worries
